File: Question8/Java/Question8.java
import java.io.*;
import java.util.*;
import java.util.stream.Stream;

class Solution {

    // You may change this function parameters
    static long countNumberOfWays(int amount, int[] coins) {
        long[] history = new long[amount + 1];
        Arrays.fill(history, 0);
        history[0] = 1;
        for (int coin : coins)
            for (int j = coin; j <= amount; j++)
                history[j] += history[j - coin];
        return history[amount];
    }


    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        String[] firstLine = scanner.nextLine().split(" ");
        String[] secondLine = scanner.nextLine().split(" ");
        scanner.close();

        int numOfUnits = Integer.parseInt(firstLine[0]);
        int numOfCoinTypes = Integer.parseInt(firstLine[1]);
        int[] coins = Stream.of(secondLine).mapToInt(Integer::parseInt).toArray();

        long result = countNumberOfWays(numOfUnits, coins);
        // Please do not remove the below line.
        System.out.println(result);
        // Do not print anything after this line
    }
}

File: Question8/Python/Question8.py
# Participants may update the following function parameters
def countNumberOfWays(numOfUnits, coins):
    history = [0] * (numOfUnits + 1)
    history[0] = 1
    for coin in range(len(coins)):
        for j in range(coins[coin], numOfUnits + 1):
            history[j] += history[j - coins[coin]]
    return history[numOfUnits]


def main():
    firstLine = input().split(" ")
    secondLine = input().split(" ")

    numOfUnits = int(firstLine[0])
    numOfCoinTypes = int(firstLine[1])
    coins = list(map(int, secondLine))

    # Participants may update the following function parameters
    answer = countNumberOfWays(numOfUnits, coins)

    # Please do not remove the below line.
    print(answer)
    # Do not print anything after this line


if __name__ == '__main__':
    main()
